### PR TITLE
Enable SystemTests failures to be processed/mass updated

### LINKED ISSUE #30273: Add mechanism to mass update system test reference files
As a developer I have made a change that requires a mass update of system test reference results without having to replace files manually.

This task is to implement a mechanism in the system test runner scripts that would allow either a fully automated or semi-automated mechanism for updating system test reference files. A bonus would be including a way of showing the differences by either plotting them or something else more relevant.

### PR DESCRIPTION
**Pull request by StephenSmith25**
**Description of work.**
This PR enables further processing of SystemTest failures. This is focussed around making it easy look at each failure by first linking each test failure to its generated mismatch workspace. With this information it is possible to create an output json file
When the validation method is WorkspaceToNexus, a full dictionary entry for the failed test is created, composing of:
```
FailedTest {
'TestName': Name of the test which failed,
'MismatchFile': Path to the mismatch file generated by the test,
'ReferenceFile': Path to reference file,
'ReferenceContentLinkFiles': Path to the original content link of the reference file
}
```

As this information provides a complete view of the file(s) required by the test, it can be used to replace the original test data using a newly generated python script addTestData.py:

`python path_to_script/Testing/SystemTests/addTestData.py --testresults FAILED-systemtests.json --replacereferences   `

This will do the following:
-	Create a new content and content link file based on the mismatch file
-	Replace the old content link file with the new one
-	Tell the user to upload the content (data) file to the test server
It can operate on multiple test failure, so can be used to mass update reference files (as long as the validation method was WorkspaceToNexus). 

Furthermore, this json file can be read in workbench to load each mismatch and reference file to compare/analyse any differences between them. 

This script can also be used to add test data, e.g if we had our new test files in a directory "testfiles":
```
/testfiles
    test1.nxs
    test2.nxs
    test3.nxs
```
 and we wanted to create our content link (hash files) in a directory ref we could do
`python path_to_script\addTestData.py --inputfiles testfiles\*.nxs --targetdirectory ref`

Which would create a content (data) file for each of the input nexus files and then create the content link (hash) file in a directory called ref
```
/ref
    test1.nxs.md5
    test2.nxs.md5
    test3.nxs.md5
```
Which you would then move into the appropriate mantid test data folder and upload the content (data) files to the Mantid data server. 

**To test:**
- Open to any feedback on this e.g. in terms of usability of the generated script
- You'll probably have to create a failing system test to test it: I was doing the following
- Open the VesuvioFittingTest system test
- Swap around the validate methods for VesuvioFittingWithQuadraticBackgroundTest and VesuvioFittingWithKFreeTest
- Build and run the system test
- Two tests will fail and a json file should be generated in the SystemTest save dir e.g. mantid_build/Testing/SystemTests/scripts:
```
{
  "FailedTests": [
    {
      "TestName": "VesuvioFittingTest.VesuvioFittingWithKFreeTest",
      "MismatchFiles": [
        "WithQuadraticBackgroundTest-mismatch.nxs"
      ],
      "ReferenceFiles": [
        "VesuvioFittingWithQuadraticBackgroundTest.nxs"
      ],
      "ReferenceContentLinkFiles": [
        "VesuvioFittingWithQuadraticBackgroundTest.nxs.md5"
      ]
    },
    {
      "TestName": "VesuvioFittingTest.VesuvioFittingWithQuadraticBackgroundTest",
      "MismatchFiles": [
        "VesuvioFittingWithKFreeTest-mismatch.nxs"
      ],
      "ReferenceFiles": [
        "VesuvioFittingWithKFreeTest.nxs"
      ],
      "ReferenceContentLinkFiles": [
        " VesuvioFittingWithKFreeTest.nxs.md5"
      ]
    }
  ]
}
```
- If you open a terminal here and do:
`python path_to_script/Testing/SystemTests/addTestData.py --testresults FAILED-systemtests.json --replacereferences   `
It should update the VesuvioFittingWithKFreeTest.nxs.md5 and VesuvioFittingWithQuadraticBackgroundTest.nxs.md5 files in the mantid source directory with a new hash (which is the hash of the mismatch workspace).

Fixes #30273 . 

*This does not require release notes* because it is an internal change.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Testing/SystemTests/scripts/runSystemTests.py [hunk 1]
```diff
@@ -346,9 +357,59 @@ def main():
...
+
+        test_mismatches = {key: status['mismatches'] for key, status in status_dict.items() if
+                           status['status'] == 'failed'}
+        output_file = output_failed_tests_information(test_mismatches, save_dir, data_paths, reference_paths)
+
         if not success:
             sys.exit(1)
 
 
+def output_failed_tests_information(test_mismatch_dict, mismatch_dir, external_data_dirs, ref_dirs):
```
Comment (robertapplin) [inline Testing/SystemTests/scripts/runSystemTests.py:369]: Would it be easy to separate this into several functions?

File: dev-docs/source/DataFilesForTesting.rst [hunk 2]
```diff
@@ -131,6 +132,22 @@ This does the following:
 -  tells the user to upload the file(s),
    ``d6948514d78db7fe251efb6cce4a9b83``, to the `remote store <https://testdata.mantidproject.org/ftp/external-data/upload>`_
 
+2: Using the python script addTestData.py in the Testing/SystemTest/scripts directory. It can be called a follows:
```
Comment (robertapplin) [inline dev-docs/source/DataFilesForTesting.rst:135]: Typo at the end of this sentence

Comment (robertapplin) [reply dev-docs/source/DataFilesForTesting.rst:135]: Maybe we should also mention that you need to be in the same directory as the `addTestData.py` script to call it this way

File: dev-docs/source/DataFilesForTesting.rst [hunk 3]
```diff
@@ -131,6 +132,22 @@ This does the following:
...
+
+.. code-block:: sh
+
+   python addTestData.py --inputfiles file1.nxs file2.nxs file3.nxs --targetdirectory Testing/Data/UnitTest/
```
Comment (robertapplin) [inline dev-docs/source/DataFilesForTesting.rst:139]: Does the `--targetdirectory` need to be a full path? If so maybe we should put `<mantid_source>/Testing/Data/UnitTest/` or something similar instead

File: dev-docs/source/SystemTests.rst [hunk 4]
```diff
@@ -360,6 +360,31 @@ by using the ``-c`` or ``--clean`` option, e.g.
 This is useful if some old data is left over from a previous run,
 where some tests were not cleanly exited.
 
+Processing test failures
```
Comment (robertapplin) [inline dev-docs/source/SystemTests.rst:363]: Nice additions to the docs 👍

File: dev-docs/source/SystemTests.rst [hunk 5]
```diff
@@ -360,6 +360,31 @@ by using the ``-c`` or ``--clean`` option, e.g.
...
+------------------------
+If any of the tests fail, a json file (FAILED-systemtests.json) containing information on the failed tests
+will be outputted. If the information was available, each entry in the file will contain the name of the failed test,
+the mismatch file generated and the name and location of the file it was compared against (the reference).
+The information can be used to automatically generate replacement references using the addTestData.py python script
```
Comment (robertapplin) [inline dev-docs/source/SystemTests.rst:368]: We could add a link to the `addTestData.py` script as is done in the release checklist to some of the helper scripts

File: dev-docs/source/SystemTests.rst [hunk 6]
```diff
@@ -360,6 +360,31 @@ by using the ``-c`` or ``--clean`` option, e.g.
...
+located in the Testing/SystemTests/scripts directory. It may be called as follows:
+
+.. code-block:: sh
+
+   python path_to_file/addTestData.py --testresults FAILED-systemtests.json --replacereferences
```
Comment (robertapplin) [inline dev-docs/source/SystemTests.rst:373]: I think `--replacereferences` should be `--replacecontentlinks`?

File: dev-docs/source/SystemTests.rst [hunk 7]
```diff
@@ -360,6 +360,31 @@ by using the ``-c`` or ``--clean`` option, e.g.
...
+
+-  computes the MD5 hash of each mismatch files
+
+-  stores each MD5 hash and replaces the old hash value with the newly generated value.
+
+-  renames the original data file to be its md5 sum
+
+-  tells the user to upload the new data files to the server.
+
+If the --replacereferences argument is omitted, it will leave the newly created hash files in the system tests
```
Comment (robertapplin) [inline dev-docs/source/SystemTests.rst:385]: Same here with `--replacereferences`

**Review by robertapplin — CHANGES_REQUESTED** [review]
This adds a very useful tool for updating system test data, good job! I only have a few comments/suggestions below.

One thought I had was whether we would want both `ReferenceFiles` and `ReferenceContentLinkFiles` in the json file (or just one of them seeing as they are almost identical and we can just add `.md5` onto the end of the `ReferenceFiles`, or the opposite way around)?

**Comment by martyngigg** [comment]
Is this something we feel we want to get into master before the code freeze or shall we tidy it up after the code freeze and put it in master? I'm inclined to go with the latter?

**Comment by martyngigg** [comment]
I'm going to move this over the 6.2 maintenance.

**Comment by martyngigg** [comment]
Could we pick this one back up this maintenance cycle? I'll move it to 6.3 as is customary for our maintenance milestones.

**Comment by StephenSmith25** [comment]
I had a look at this again last maintenance period and it looked like the conflicts in xmlreporter and runSystemTests broke my previous solution. Unfortunately I won't have time to finish this off.

**Pull request closed without merging**